File: go/fixture/go-mini/internal/snapshot/schedule.go
package snapshot

import (
	"errors"
	"fmt"
	"slices"
	"time"
)

const minutesPerDay = 24 * 60

// Window is a daily time-of-day interval during which snapshots may be taken.
// A window that ends before it starts wraps past midnight.
type Window struct {
	start int // minutes since midnight
	end   int // minutes since midnight
}

// ParseWindow builds a Window from two "HH:MM" clock times.
func ParseWindow(start, end string) (Window, error) {
	s, err := parseClock(start)
	if err != nil {
		return Window{}, fmt.Errorf("window start: %w", err)
	}
	e, err := parseClock(end)
	if err != nil {
		return Window{}, fmt.Errorf("window end: %w", err)
	}
	if s == e {
		return Window{}, fmt.Errorf("window %s-%s is empty", start, end)
	}
	return Window{start: s, end: e}, nil
}

func parseClock(v string) (int, error) {
	t, err := time.Parse("15:04", v)
	if err != nil {
		return 0, fmt.Errorf("clock time %q: want HH:MM", v)
	}
	return t.Hour()*60 + t.Minute(), nil
}

// Contains reports whether t falls inside the window.
func (w Window) Contains(t time.Time) bool {
	m := t.Hour()*60 + t.Minute()
	if w.start < w.end {
		return m >= w.start && m < w.end
	}
	return m >= w.start || m < w.end
}

// Duration returns how long the window stays open each day.
func (w Window) Duration() time.Duration {
	span := w.end - w.start
	if span < 0 {
		span += minutesPerDay
	}
	return time.Duration(span) * time.Minute
}

// String renders the window as "HH:MM-HH:MM".
func (w Window) String() string {
	return fmt.Sprintf("%02d:%02d-%02d:%02d", w.start/60, w.start%60, w.end/60, w.end%60)
}

// Plan is the set of windows during which a device may be snapshotted.
type Plan struct {
	windows []Window
}

// NewPlan builds a Plan from at least one window. The plan keeps its own copy
// so later changes to the caller's slice cannot reorder or shrink it.
func NewPlan(windows []Window) (Plan, error) {
	if len(windows) == 0 {
		return Plan{}, errors.New("plan: no windows")
	}
	return Plan{windows: slices.Clone(windows)}, nil
}

// Open reports whether any window of the plan contains t.
func (p Plan) Open(t time.Time) bool {
	return slices.ContainsFunc(p.windows, func(w Window) bool { return w.Contains(t) })
}

// Windows returns the plan's windows. Callers get their own copy: the plan is
// shared by the scheduler and the status page, and neither may edit it.
func (p Plan) Windows() []Window {
	return slices.Clone(p.windows)
}

File: py/fixture/py-mini/internal/snapshot/schedule.py
"""Daily windows during which snapshots may be taken."""

import time
from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Self

_MINUTES_PER_DAY = 24 * 60


@dataclass(frozen=True)
class Window:
    """A daily time-of-day interval during which snapshots may be taken.

    A window that ends before it starts wraps past midnight. Both ends are
    minutes since midnight and are checked as the Window is built, so a
    Window in circulation is never empty and never points outside the day.
    """

    start: int  # minutes since midnight
    end: int  # minutes since midnight

    def __post_init__(self) -> None:
        for name, minutes in (("start", self.start), ("end", self.end)):
            if not 0 <= minutes < _MINUTES_PER_DAY:
                raise ValueError(f"window {name}: {minutes} minutes is outside the day")
        if self.start == self.end:
            raise ValueError(f"window {self} is empty")

    @classmethod
    def parse(cls, start: str, end: str) -> Self:
        """Build a Window from two "HH:MM" clock times."""
        try:
            s = _parse_clock(start)
        except ValueError as err:
            raise ValueError(f"window start: {err}") from err
        try:
            e = _parse_clock(end)
        except ValueError as err:
            raise ValueError(f"window end: {err}") from err
        return cls(start=s, end=e)

    def contains(self, t: datetime) -> bool:
        """Report whether t falls inside the window."""
        m = t.hour * 60 + t.minute
        if self.start < self.end:
            return self.start <= m < self.end
        return m >= self.start or m < self.end

    def duration(self) -> timedelta:
        """Return how long the window stays open each day."""
        span = self.end - self.start
        if span < 0:
            span += _MINUTES_PER_DAY
        return timedelta(minutes=span)

    def __str__(self) -> str:
        return (
            f"{self.start // 60:02d}:{self.start % 60:02d}-{self.end // 60:02d}:{self.end % 60:02d}"
        )


def _parse_clock(v: str) -> int:
    try:
        t = time.strptime(v, "%H:%M")
    except ValueError:
        raise ValueError(f'clock time "{v}": want HH:MM') from None
    return t.tm_hour * 60 + t.tm_min


class Plan:
    """The set of windows during which a device may be snapshotted."""

    def __init__(self, windows: list[Window]) -> None:
        """Build a Plan from at least one window.

        The plan keeps its own copy so later changes to the caller's list
        cannot reorder or shrink it.
        """
        if not windows:
            raise ValueError("plan: no windows")
        self._windows = list(windows)

    def open(self, t: datetime) -> bool:
        """Report whether any window of the plan contains t."""
        return any(w.contains(t) for w in self._windows)

    def windows(self) -> list[Window]:
        """Return the plan's windows.

        Callers get their own copy: the plan is shared by the scheduler and
        the status page, and neither may edit it.
        """
        return list(self._windows)
